pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
         )
         results = self._inner_execute(jac_tapes)
         return tuple(batch_post_processing(results))
+
+
+class DeviceDerivatives(JacobianProductCalculator):
```

Comment:
Not necessarily in this PR: but how do we choose between the methods in the interfaces? Let's say you can have all three available jacobian available, vjp with multiplication/without multiplication and jvp with multiplication/without multiplication?

Reply:
Autograd and torch will use `compute_vjp`. 

Tensorflow will sometimes use `compute_vjp` and sometimes use `compute_jacobian`, based on whether or not it wants to trace the calculation.

Jax will use `execute_and_compute_jvp` and `compute_vjp`.  

Jax-jit will use `compute_jacobian`.

Reply:
This hypothetical but let's imagine a device developer comes with a device that can only compute the jacobians (like default qubit legacy for example), how do we make sure that it will work with our interfaces/execution. I think `compute_jacobian` should be the fallback for all devices. Also a user might want to choose between jacobians and vjp/jvps. Will there be a way to toggle between the methods?

Reply:
Yep.  See PR #4557: https://github.com/PennyLaneAI/pennylane/pull/4557

A device under the new interface will let pennylane know it prefers to use the device jacobian product by filling in `ExecutionConfig.use_device_jacobian_product = True`. 

https://github.com/PennyLaneAI/pennylane/blob/633c1c4cba801f408163361cec88a102a40881b7/pennylane/devices/qubit/preprocess.py#L415

When `qml.execute` sees that option is `True`, it will create a `DeviceJacobianProducts` instance.  Otherwise, it will create a `DeviceDerivatives` instance.

https://github.com/PennyLaneAI/pennylane/blob/633c1c4cba801f408163361cec88a102a40881b7/pennylane/interfaces/execution.py#L644

```
    if config.use_device_jacobian_product:
        jpc = DeviceJacobianProducts(device, config)

    elif config.use_device_gradient:
        jpc = DeviceDerivatives(device, config)
```

In the short term, I propose to add a keyword argument `use_device_jacobian_product` to first `qml.execute` and then `QNode`.  If `True`, then we will either use the device provided jacobian product or fail.  If `None`, we will query the device for what it wants to do and do that.  If `False`, we will use use the full jacobian.

https://github.com/PennyLaneAI/pennylane/blob/633c1c4cba801f408163361cec88a102a40881b7/pennylane/interfaces/execution.py#L414

We can start with the argument as `False`, and then switch it to `None` once we know the system is working well.


Lots of steps, but the guiding principles are:

* Interface code only pays attention to things that differ between interfaces.  `autograd.py` shouldn't care about the specifics of how pennylane computes the vjp.  It just should know how to bind a vjp to autograd.  This makes extending how we compute vjps easier.


* Devices should be able to configure what they deem "best".  By filling in things like `config.use_device_jacobian_product`, `config.use_device_gradient`, and `config.grad_on_execution`, a device tells pennylane what to do with it.


Hope that clears things up a little bit.

Reply:
Thanks it looks good to me :+1: